What's new in ASP.NET Core 5 (#20140)
dotnet/AspNetCore.Docs: aspnetcore/release-notes/aspnetcore-5.0.md

```xml
<ItemGroup>
    <PackageReference Include="Swashbuckle.AspNetCore" Version="5.5.1" />
</ItemGroup>
```

The template generated code contains code in `Startup.ConfigureServices` that activates OpenAPI document generation:

[!code-csharp[](~/release-notes/sample/StartupSwagger.cs?name=snippet)]

The `Startup.Configure` method adds the Swashbuckle middleware, which enables the:

* Document generation process.
* Swagger UI page by default in development mode.

The template generated code won't accidentally expose the API's description when publishing to production.

[!code-csharp[](~/release-notes/sample/StartupSwagger.cs?name=snippet2)]

#### Azure API Management Import

When ASP.NET Core API projects enable OpenAPI, the Visual Studio 2019 version 16.8 and later publishing automatically offer an additional step in the publishing flow. Developers who use [Azure API Management](xref:tutorials/publish-to-azure-api-management-using-vs) have an opportunity to automatically import the APIs into Azure API Management during the publish flow:

![Azure API Management Import VS publishing](~/release-notes/static/publish-to-apim.png)

### Better launch experience for web API projects

> With OpenAPI enabled by default, the app launching experience (F5) for web API developers significantly improves. With .NET, the web API template comes pre-configured to load up the Swagger UI page. The Swagger UI page provides both the documentation added for the published API, and enables testing the APIs with a single click.

Reviewer: ```suggestion
With OpenAPI enabled by default, the app launching experience (F5) for web API developers significantly improves. With ASP.NET Core 5.0, the web API template comes pre-configured to load up the Swagger UI page. The Swagger UI page provides both the documentation added for the published API, and enables testing the APIs with a single click.
```

![swagger/index.html view](~/release-notes/static/swagger-ui-page-rc1.png)

> ## Blazor

Reviewer 2: We did a bunch of work to significantly improve runtime performance in Blazor WebAssembly for .NET 5. Can we add a section to highlight that? Details here: https://devblogs.microsoft.com/aspnet/asp-net-core-updates-in-net-5-release-candidate-1/#blazor-webassembly-performance-improvements

Custom validation class names are useful when integrating with CSS frameworks, such as Bootstrap. For more information, see <xref:blazor/forms-validation#custom-validation-class-attributes>.

### IAsyncDisposable support

Blazor components now support the <xref:System.IAsyncDisposable> interface for the asynchronous release of allocated resources.

### JavaScript isolation and object references

Blazor enables JavaScript isolation in standard [JavaScript modules](https://developer.mozilla.org/docs/Web/JavaScript/Guide/Modules). For more information, see <xref:blazor/call-javascript-from-dotnet#blazor-javascript-isolation-and-object-references>.

### Form components support display name

The following built-in components support display names with the `DisplayName` parameter:

* `InputDate`
* `InputNumber`
* `InputSelect`

For more information, see <xref:blazor/forms-validation#display-name-support>.

### Catch-all route parameters

Catch-all route parameters, which capture paths across multiple folder boundaries, are supported in components. For more information, see <xref:blazor/fundamentals/routing#catch-all-route-parameters>.

### Debugging improvements

> Debugging Blazor WebAssembly apps is improved in ASP.NET Core 5.0. Additiionally, debugging is now supported on Visual Studio for Mac. For more information, see <xref:blazor/debug>.

Reviewer 3: ```suggestion
Debugging Blazor WebAssembly apps is improved in ASP.NET Core 5.0. Additionally, debugging is now supported on Visual Studio for Mac. For more information, see <xref:blazor/debug>.
```

### Microsoft Identity v2.0 and MSAL v2.0

Blazor security now uses Microsoft Identity v2.0 ([`Microsoft.Identity.Web`](https://www.nuget.org/packages/Microsoft.Identity.Web) and [`Microsoft.Identity.Web.UI`](https://www.nuget.org/packages/Microsoft.Identity.Web.UI)) and MSAL v2.0. For more information, see the topics in the [Blazor Security and Identity node](xref:blazor/security/index).

### State management support

> `Microsoft.AspNetCore.ProtectedBrowserStorage` has been moved to `Microsoft.AspNetCore.Components.Web.Extensions` and has undergone API changes. State management documentation has been split across Blazor hosting models. For more information, see <xref:blazor/state-management>.

Reviewer 2: We ended up adding the protected browser storage APIs into the ASP.NET Core shared framework. No need to add any additional packages. This API makes it easy to securely store app state client-side in the browser for Blazor Server apps.

<!-- WAIT ON THIS ENTRY UNTIL https://github.com/dotnet/AspNetCore.Docs/pull/19887 MERGES.

> ### Component integration across hosting models

Reviewer 2: This same functionality also enabled prerendering of Blazor WebAssembly apps. I think Blazor WebAssembly prerendering is signficant enough that it should have its own section.

Component integration is improved across hosting models. For more information, see <xref:blazor/components/integrate-components-into-razor-pages-and-mvc-apps>.

-->

> ### Linker is now a Trimmer

Reviewer 2: Functionally we do the same thing in both Blazor WebAssembly 3.2 and in .NET 5: we trim out any unused code from your published app to reduce its download size. Whether you call it "linking" or "trimming" the effect is the same. In .NET 5 there are a bunch of improvements in IL trimming that Blazor takes advantage of. Presumably that should be the focus of this section, not the change in verbiage.

> ### Protected Browser Support API
> 
> The Protected Browser Support API for Blazor Server has moved into the shared framework. You no longer need to add a separate package to the app, and the services are registered by the framework. For more information, see <xref:blazor/state-management>.

Reviewer 2: ```suggestion
### Protected Browser Storage for Blazor Server

Blazor Server apps can now use built-in support for storing app state in the browser that has been protected using ASP.NET Core data protection. Data can be stored in either local browser storage or session storage. For more information, see <xref:blazor/state-management>.
```

> Protected browser storage APIs (`Microsoft.AspNetCore.ProtectedBrowserStorage`) have been moved to `Microsoft.AspNetCore.Components.Web.Extensions` in the shared framework and have undergone API changes. State management documentation has been split across Blazor hosting models. For more information, see <xref:blazor/state-management>.

Reviewer 2: This section seems to be talking about preview-to-preview changes, but this feature is really a new feature in .NET 5 (previously it was an experimental package).

The protected browser support was also moved into the shared framework in RC2, so you no longer need to add a separate package and the services are registered when you add the services for Blazor Server. You just need to add a `using` directive for `Microsoft.AspNetCore.Componets.Server.ProtectedBrowserStorage`. This needs to be updated in the Blazor docs too: https://docs.microsoft.com/en-us/aspnet/core/blazor/state-management?view=aspnetcore-5.0&pivots=server#aspnet-core-protected-browser-storage.

For more information, see [Use hub filters in ASP.NET Core SignalR](xref:signalr/hub-filters).

### SignalR parallel hub invocations

ASP.NET Core SignalR is now capable of handling parallel hub invocations. The default behavior can be changed to allow clients to invoke more than one hub method at a time:

[!code-csharp[](~/release-notes/sample/StartupSignalRhubs.cs?name=snippet)]

### Added Messagepack support in SignalR Java client

A new package, [com.microsoft.signalr.messagepack](https://mvnrepository.com/artifact/com.microsoft.signalr.messagepack), adds MessagePack support to the SignalR Java client. To use the MessagePack hub protocol, add `.withHubProtocol(new MessagePackHubProtocol())` to the connection builder:

```java
HubConnection hubConnection = HubConnectionBuilder.create(
                           "http://localhost:53353/MyHub")
               .withHubProtocol(new MessagePackHubProtocol())
               .build();
```

<!--
See [Update SignalR code](xref:migration/31-to-50#signalr) for migration instructions.
-->

## Kestrel

> * Reloadable endpoints via configuration: Kestrel can detect changes to configuration passed to [KestrelServerOptions.Configure](xref:Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServerOptions.Configure%2A) and unbind from existing endpoints and bind to new endpoints without requiring an application restart. By default when using <xref:Microsoft.Extensions.Hosting.GenericHostBuilderExtensions.ConfigureWebHostDefaults%2A> or <xref:Microsoft.AspNetCore.WebHost.CreateDefaultBuilder%2A>, Kestrel binds to the "Kestrel" configuration subsection.

Reviewer 4: I think think "Kestrel" should probably be left as a string literal. It's referencing the string literal passed to `IConfiguration.GetSection("Kestrel")` [here](https://github.com/dotnet/aspnetcore/blob/7e9e03b70124784b1de5564c573bd65cdaccbfcc/src/DefaultBuilder/src/WebHost.cs#L226).

We should point out that you need to pass `reloadOnChange: true` when calling `KestrelServerOptions.Configure` manually to get reloadable endpoints though.

```suggestion
* Reloadable endpoints via configuration: Kestrel can detect changes to configuration passed to [KestrelServerOptions.Configure](xref:Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServerOptions.Configure%2A) and unbind from existing endpoints and bind to new endpoints without requiring an application restart when the `reloadOnChange` parameter is `true`. By default when using <xref:Microsoft.Extensions.Hosting.GenericHostBuilderExtensions.ConfigureWebHostDefaults%2A> or <xref:Microsoft.AspNetCore.WebHost.CreateDefaultBuilder%2A>, Kestrel binds to the "Kestrel" configuration subsection with `reloadOnChange` enabled.
```

> * Reloadable endpoints via configuration: Kestrel can detect changes to configuration passed to [KestrelServerOptions.Configure](xref:Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServerOptions.Configure%2A) and unbind from existing endpoints and bind to new endpoints without requiring an application restart. By default when using <xref:Microsoft.Extensions.Hosting.GenericHostBuilderExtensions.ConfigureWebHostDefaults%2A> or <xref:Microsoft.AspNetCore.WebHost.CreateDefaultBuilder%2A>, Kestrel binds to the `Kestrel` configuration subsection.

Reviewer 4: I originally made this comment at https://github.com/dotnet/AspNetCore.Docs/pull/20140#discussion_r514507772, but that conversation was resolved and I don't want this to be missed.

I think think "Kestrel" should probably be left as a string literal. It's referencing the string literal passed to `IConfiguration.GetSection("Kestrel")` [here](https://github.com/dotnet/aspnetcore/blob/7e9e03b70124784b1de5564c573bd65cdaccbfcc/src/DefaultBuilder/src/WebHost.cs#L226).

We should point out that you need to pass `reloadOnChange: true` when calling `KestrelServerOptions.Configure` manually to get reloadable endpoints though.

```suggestion
* Reloadable endpoints via configuration: Kestrel can detect changes to configuration passed to [KestrelServerOptions.Configure](xref:Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServerOptions.Configure%2A) and unbind from existing endpoints and bind to new endpoints without requiring an application restart when the `reloadOnChange` parameter is `true`. By default when using <xref:Microsoft.Extensions.Hosting.GenericHostBuilderExtensions.ConfigureWebHostDefaults%2A> or <xref:Microsoft.AspNetCore.WebHost.CreateDefaultBuilder%2A>, Kestrel binds to the "Kestrel" configuration subsection with `reloadOnChange` enabled.
```

The template generated code won't accidentally expose the API's description when publishing to production.

[!code-csharp[](~/release-notes/sample/StartupSwagger.cs?name=snippet2)]

#### Azure API Management Import

When ASP.NET Core API projects enable OpenAPI, the Visual Studio 2019 version 16.8 Preview 3.2 and later publishing automatically offer an additional step in the publishing flow. Developers who use [Azure API Management](xref:tutorials/publish-to-azure-api-management-using-vs) have an opportunity to automatically import the APIs into Azure API Management during the publish flow:

![Azure API Management Import VS publishing](~/release-notes/static/publish-to-apim.png)

### Better launch experience for web API projects

With OpenAPI enabled by default, the app launching experience (F5) for web API developers significantly improves. With .NET, the web API template comes pre-configured to load up the Swagger UI page. The Swagger UI page provides both the documentation added for the published API, and enables testing the APIs with a single click.

![swagger/index.html view](~/release-notes/static/swagger-ui-page-rc1.png)

## Blazor

> ### `JSObjectReference` replacement

Reviewer 2: This was a preview-to-preview change within the .NET 5 release and not something that needs to be highlighted here. 

Instead, we should add sections for JavaScript isolation via JavaScript modules and capturing JS object references via `IJSObjectReference`. See https://docs.microsoft.com/aspnet/core/blazor/call-javascript-from-dotnet?view=aspnetcore-5.0#blazor-javascript-isolation-and-object-references for details.

Reviewer 2: Looks like you cover JS isolation and object refs below, so I think this section can just be removed.

`JSObjectReference` has been made internal, and `IJSObjectReference` takes its place in the public API. For more information, see the [official announcement (aspnet/Announcements #435)](https://github.com/aspnet/Announcements/issues/435) and <xref:blazor/call-javascript-from-dotnet?view=aspnetcore-5.0#blazor-javascript-isolation-and-object-references>.

> ### CSS isolation

Reviewer 2: Add a link to the corresponding documentation: https://docs.microsoft.com/aspnet/core/blazor/components/css-isolation?view=aspnetcore-5.0

Blazor now supports defining CSS styles that are scoped to a given component. Component specific CSS styles make it easier to reason about the styles in an app and to avoid unintentional side effects from global styles. <!-- UNCOMMENT CROSS-LINK AFTER https://github.com/dotnet/AspNetCore.Docs/pull/19956 MERGES: For more information, see <xref:blazor/components/css-isolation>.-->

### New `InputFile` component

The `InputFile` component allows reading one or more files selected by a user to be upload. For more information, see <xref:blazor/file-uploads>.

### New `InputRadio` and `InputRadioGroup` components

> Blazor has built-in `InputRadio` and `InputRadioGroup` components that simplify data binding to radio button groups with integrated validation. For more information, see <xref:blazor/forms-validation>.

Reviewer 2: Can we link directly to https://docs.microsoft.com/aspnet/core/blazor/forms-validation?view=aspnetcore-5.0#radio-buttons ?

### `ontoggle` event support

Blazor events now support the `ontoggle` DOM event. For more information, see <xref:blazor/components/event-handling#event-argument-types>.

### Set UI focus in Blazor apps

Use the `FocusAsync` convenience method on element references for setting the UI focus on that element.
<!-- UNCOMMENT CROSS-LINK WHEN https://github.com/dotnet/AspNetCore.Docs/issues/20232 IS WORKED. For more information, see <xref:blazor/components/event-handling#{ADD BOOKMARK}>.-->

### Custom validation class attributes

Custom validation class names are useful when integrating with CSS frameworks, such as Bootstrap. For more information, see <xref:blazor/forms-validation#custom-validation-class-attributes>.

### IAsyncDisposable support

Blazor components now support the <xref:System.IAsyncDisposable> interface for the asynchronous release of allocated resources.

### JavaScript isolation and object references

Blazor enables JavaScript isolation in standard [JavaScript modules](https://developer.mozilla.org/docs/Web/JavaScript/Guide/Modules). For more information, see <xref:blazor/call-javascript-from-dotnet#blazor-javascript-isolation-and-object-references>.

> ### Influence the HTML head in Blazor apps

Reviewer 2: These components didn't end up getting finalized for .NET 5, so this section should be removed.

Reviewer 5: @danroth27 ... We placed 5.0 content into the topics, so I'll comment it out in-place on a new PR. It was reviewed and is good coverage, so we'll have it in-place for 5.x later.

### Browser compatibility analyzer

Blazor WebAssembly apps target the full .NET API surface area, but not all .NET APIs are supported on WebAssembly due to browser sandbox constraints. Unsupported APIs throw <xref:System.PlatformNotSupportedException> when running on WebAssembly. A platform compatibility analyzer warns the developer when the app uses APIs that aren't supported by the app's target platforms. For more information, see <xref:blazor/components/class-libraries#browser-compatibility-analyzer-for-blazor-webassembly>.

### Lazy load assemblies

Blazor WebAssembly app startup performance can be improved by deferring the loading of some application assemblies until they are required. For more information, see <xref:blazor/webassembly-lazy-load-assemblies>.

### Updated globalization support

Globalization support is available for Blazor WebAssembly based on International Components for Unicode (ICU). For more information, see <xref:blazor/globalization-localization>.

## gRPC

> Many preformance improvements have been made in [gRPC](https://grpc.io/). <!-- For more information, see [this ]

Reviewer 2: gRPC perf blog post is now live: https://devblogs.microsoft.com/aspnet/grpc-performance-improvements-in-net-5/

For more information, see [Use hub filters in ASP.NET Core SignalR](xref:signalr/hub-filters).

### SignalR parallel hub invocations

ASP.NET Core SignalR is now capable of handling parallel hub invocations. The default behavior can be changed to allow clients to invoke more than one hub method at a time:

[!code-csharp[](~/release-notes/sample/StartupSignalRhubs.cs?name=snippet)]

### Added Messagepack support in SignalR Java client

A new package, [com.microsoft.signalr.messagepack](https://search.maven.org/search?q=g:com.microsoft.signalr.messagepack), adds MessagePack support to the SignalR Java client. To use the MessagePack hub protocol, add `.withHubProtocol(new MessagePackHubProtocol())` to the connection builder:

```java
HubConnection hubConnection = HubConnectionBuilder.create(
                           "http://localhost:53353/MyHub")
               .withHubProtocol(new MessagePackHubProtocol())
               .build();
```

<!--
See [Update SignalR code](xref:migration/31-to-50#signalr) for migration instructions.
-->

## Kestrel

> * Reloadable endpoints via configuration: Kestrel can detect changes to configuration passed to [KestrelServerOptions.Configure](xref:Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServerOptions.Configure%2A) and unbind from existing endpoints and bind to new endpoints without requiring an application restart.

Reviewer 4: I think we should point out that the "Kestrel" configuration subsection is bound by default when you call `.ConfigureWebHostDefaults(...` like our default templates do. Maybe something like the following.

```suggestion
* Reloadable endpoints via configuration: Kestrel can detect changes to configuration passed to [KestrelServerOptions.Configure](xref:Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServerOptions.Configure%2A) and unbind from existing endpoints and bind to new endpoints without requiring an application restart. By default when using <xref:Microsoft.Extensions.Hosting.GenericHostBuilderExtensions.ConfigureWebHostDefaults%2A> or <xref:Microsoft.AspNetCore.WebHost.CreateDefaultBuilder%2A>, Kestrel binds to the "Kestrel" configuration subsection.
```

* HTTP/2 response headers improvements. For more information, see [Performance improvements](#performance-improvements) in the next section.
* Support for additional endpoints types in the sockets transport: Adding to the new API introduced in <xref:System.Net.Sockets?displayProperty=nameWithType>, the sockets default transport in [Kestrel](xref:fundamentals/servers/kestrel) allows binding to both existing file handles and Unix domain sockets. Support for binding to existing file handles enables using the existing `Systemd` integration without requiring the `libuv` transport.

> * Custom header decoding in Kestrel: Apps can specify which <xref:System.Text.Encoding> to use to interpret incoming headers based on the header name instead of defaulting to `UTF-8`. Set the <xref:System.Net.Http.SocketsHttpHandler.RequestHeaderEncodingSelector> property on <xref:Microsoft.AspNetCore.Server.Kestrel.KestrelServerOptions> to specify which encoding to use:

Reviewer 4: Kestrel's selector is a `Func<string, Encoding>` while SocketsHttpHandler's selector is a `HeaderEncodingSelector<System.Net.Http.HttpRequestMessage>`.

```suggestion
* Custom header decoding in Kestrel: Apps can specify which <xref:System.Text.Encoding> to use to interpret incoming headers based on the header name instead of defaulting to `UTF-8`. Set the <xref:Microsoft.AspNetCore.Server.Kestrel.KestrelServerOptions.RequestHeaderEncodingSelector> property to specify which encoding to use:
```

---
title: What's new in ASP.NET Core 5.0
author: rick-anderson
description: Learn about the new features in ASP.NET Core 5.0.
ms.author: riande
ms.custom: mvc
ms.date: 12/05/2019
no-loc: ["ASP.NET Core Identity", cookie, Cookie, Blazor, "Blazor Server", "Blazor WebAssembly", "Identity", "Let's Encrypt", Razor, SignalR]
uid: aspnetcore-5.0
---
# What's new in ASP.NET Core 5.0

This article highlights the most significant changes in ASP.NET Core 5.0 with links to relevant documentation.

## ASP.NET Core MVC and Razor improvements

### Model binding DateTime as UTC

> Model binding now supports binding UTC time strings to `DateTime`. If the request contains a UTC time string, model binding binds it to a UTC `DateTime`. For example, the following time string is bound the UTC `DateTime`:  `https://example.com/mycontroller/myaction?time=2019-06-14T02%3A30%3A04.0576719Z`

Reviewer: ```suggestion
Model binding now supports binding UTC time strings to `DateTime`. If the request contains a UTC time string, model binding binds it to a UTC `DateTime`. For example, the following time string is bound the UTC `DateTime`: `https://example.com/mycontroller/myaction?time=2019-06-14T02%3A30%3A04.0576719Z`
```

### Model binding and validation with C# 9 record types

[C# 9 record types](/dotnet/csharp/whats-new/csharp-9#record-types) can be used with model binding in an MVC controller or a Razor Page. Record types are a good way to model data being transmitted over the network.

For example, the following `PersonController` uses the `Person` record type with model binding and form validation:

```csharp
public record Person([Required] string Name, [Range(0, 150)] int Age);

public class PersonController
{
   public IActionResult Index() => View();

   [HttpPost]
   public IActionResult Index(Person person)
   {
          // ...
   }
}
```

The `Person/Index.cshtml` file:

```cshtml
@model Person

Name: <input asp-for="Model.Name" />
<span asp-validation-for="Model.Name" />

Age: <input asp-for="Model.Age" />
<span asp-validation-for="Model.Age" />
```

### Improvements to DynamicRouteValueTransformer

> ASP.NET Core 3.1 introduced <xref:Microsoft.AspNetCore.Mvc.Routing.DynamicRouteValueTransformer> as a way to use use a custom endpoint to dynamically select an MVC controller action or a razor page.  ASP.NET Core 5.0 apps can pass state to a `DynamicRouteValueTransformer` and filter the set of endpoints chosen.

Reviewer: ```suggestion
ASP.NET Core 3.1 introduced <xref:Microsoft.AspNetCore.Mvc.Routing.DynamicRouteValueTransformer> as a way to use use a custom endpoint to dynamically select an MVC controller action or a Razor page. ASP.NET Core 5.0 apps can pass state to a `DynamicRouteValueTransformer` and filter the set of endpoints chosen.
```

The validation system in .NET Core 3.0 and later treats non-nullable parameters or bound properties as if they had a [Required] attribute.
see https://docs.microsoft.com/aspnet/core/mvc/models/validation?view=aspnetcore-3.1   
-->

## Web API

### OpenAPI Specification on by default

[OpenAPI Specification](http://spec.openapis.org/oas/v3.0.3) is a industry adopted convention for describing HTTP APIs and integrating them into complex business processes or with 3rd parties. OpenAPI is widely supported by all cloud providers and many API registries. Apps that emit OpenAPI documents from Web APIs have a variety of new opportunities in which those APIs can be used. In partnership with the maintainers of the open-source project [Swashbuckle.AspNetCore](https://www.nuget.org/packages/Swashbuckle.AspNetCore/), the ASP.NET Core API template contains a NuGet dependency on [Swashbuckle](https://github.com/domaindrivendev/Swashbuckle.AspNetCore), a popular open-source NuGet package that emits OpenAPI documents dynamically. Swashbuckle does this by introspecting over the API Controllers and generating the OpenAPI document at run-time, or at build time using the Swashbuckle CLI.

> In .NET 5, the Web API templates enable the OpenAPI output being enabled by default. To disable OpenAPI:

Reviewer: ```suggestion
In ASP.NET Core 5.0, the web API templates enable the OpenAPI support by default. To disable OpenAPI:
```

> * From the command line: `dotnet new webapi --no-openapi true`

Reviewer: ```suggestion
* From the command line:

	```dotnetcli
	dotnet new webapi --no-openapi true`
	```
```

* From Visual Studio: Uncheck **Enable OpenAPI support**.

> All *.csproj* files created for Web API projects contain the [Swashbuckle.AspNetCore](https://www.nuget.org/packages/Swashbuckle.AspNetCore/) NuGet package reference.

Reviewer: ```suggestion
All *.csproj* files created for web API projects contain the [Swashbuckle.AspNetCore](https://www.nuget.org/packages/Swashbuckle.AspNetCore/) NuGet package reference.
```

```xml
<ItemGroup>
    <PackageReference Include="Swashbuckle.AspNetCore" Version="5.5.1" />
</ItemGroup>
```

The template generated code contains code in `Startup.ConfigureServices` that activates OpenAPI document generation:

[!code-csharp[](~/release-notes/sample/StartupSwagger.cs?name=snippet)]

The `Configure` method adds the Swashbuckle middleware, which enables the:

* Document generation process.
* Swagger UI page by default in development mode.

> The template generated code won't accidentally expose the API’s description when publishing to production.

Reviewer: ```suggestion
The template generated code won't accidentally expose the API's description when publishing to production.
```

[!code-csharp[](~/release-notes/sample/StartupSwagger.cs?name=snippet2)]

#### Azure API Management Import

When ASP.NET Core API projects enable OpenAPI, the Visual Studio 2019 version 16.8 Preview 3.2 and later publishing automatically offer an additional step in the publishing flow. Developers who use [Azure API Management](xref:tutorials/publish-to-azure-api-management-using-vs) have an opportunity to automatically import the APIs into Azure API Management during the publish flow:

![Azure API Management Import VS publishing](~/release-notes/static/publish-to-apim.png)

> ### Better F5 Experience for Web API Projects

Reviewer: Since not everyone uses F5 to start the app, we should use a different term. How about "start" or "launch"?

## gRPC

Many preformance improvements have been made in [gRPC](https://grpc.io/). <!-- For more information, see [this ]
Waiting for https://github.com/aspnet/specs/pull/523
--> 
For more information, see <xref:grpc/index>.

## SignalR

SignalR Hub filters, called Hub pipelines in ASP.NET SignalR, is a feature that allows code to run before and after Hub methods are called. Running code before and after Hub methods are called is similar to how middleware has the ability to run code before and after an HTTP request. Common uses include logging, error handling, and argument validation.

For more information, see [Use hub filters in ASP.NET Core SignalR](xref:signalr/hub-filters).

### SignalR parallel hub invocations

ASP.NET Core SignalR is now capable of handling parallel hub invocations. The default behavior can be changed to allow clients to invoke more than one hub method at a time:

[!code-csharp[](~/release-notes/sample/StartupSignalRhubs.cs?name=snippet)]

### Added Messagepack support in SignalR Java client

<!-- Review: I can't find NuGet com.microsoft.signalr.messagepack  -->

> A new package, `com.microsoft.signalr.messagepack`, adds messagepack support to the SignalR java client. To use the messagepack hub protocol, add `.withHubProtocol(new MessagePackHubProtocol())` to the connection builder:

Reviewer: ```suggestion
A new package, `com.microsoft.signalr.messagepack`, adds MessagePack support to the SignalR Java client. To use the MessagePack hub protocol, add `.withHubProtocol(new MessagePackHubProtocol())` to the connection builder:
```

> ```csharp

Reviewer: ```suggestion
```java
```

HubConnection hubConnection = HubConnectionBuilder.create(
                           "http://localhost:53353/MyHub")
               .withHubProtocol(new MessagePackHubProtocol())
               .build();

>                

Reviewer: ```suggestion
```

```

<!--
See [Update SignalR code](xref:migration/31-to-50#signalr) for migration instructions.
-->

## Kestrel

* Reloadable endpoints via configuration: Kestrel can detect changes to configuration passed to [KestrelServerOptions.Configure](xref:Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServerOptions.Configure%2A) and unbind from existing endpoints and bind to new endpoints without requiring an application restart.
* HTTP/2 response headers improvements. For more information, see [Performance improvements](#performance-improvements) in the next section.

> * Support for additional endpoints types in the sockets transport: Adding to the new API introduced in <xref:System.Net.Sockets?displayProperty=nameWithType>, the sockets default transport in [Kestrel](xref:fundamentals/servers/kestrel) allows binding to both existing file handles and unix domain sockets. Support for binding to existing file handles enables using the existing `Systemd` integration without requiring the `libuv` transport.

Reviewer: ```suggestion
* Support for additional endpoints types in the sockets transport: Adding to the new API introduced in <xref:System.Net.Sockets?displayProperty=nameWithType>, the sockets default transport in [Kestrel](xref:fundamentals/servers/kestrel) allows binding to both existing file handles and Unix domain sockets. Support for binding to existing file handles enables using the existing `Systemd` integration without requiring the `libuv` transport.
```

            {
                options.RequestHeaderEncodingSelector = encoding =>
                {
                    switch (encoding)
                    {
                        case "Host":
                            return System.Text.Encoding.Latin1;
                        default:
                            return System.Text.Encoding.UTF8;
                    }
                };
            });
            webBuilder.UseStartup<Startup>();
        });
  ```

### Kestrel endpoint-specific options via configuration

Support has been added for configuring Kestrel’s endpoint-specific options via [configuration](xref:fundamentals/configuration/index). The endpoint-specific configurations includes the:

* Http protocols used
* TLS protocols used
* Certificate selected
* Cient certificate mode

Configuration allows specifying which certificate is selected based on the specified server name. The server name is part of the Server Name Indication (SNI) extension to the TLS protocol as indicated by the client. Kestrel’s configuration also supports a wildcard prefix in the host name.

The following example shows how to specify endpoint-specific using a configuration file:

> ```xml

Reviewer: ```suggestion
```json
```

<xref:Microsoft.AspNetCore.Server.Kestrel.KestrelServerOptions>:

   ```csharp
   public static IHostBuilder CreateHostBuilder(string[] args) =>
    Host.CreateDefaultBuilder(args)
        .ConfigureWebHostDefaults(webBuilder =>
        {
            webBuilder.ConfigureKestrel(options =>
            {
                options.Limits.Http2.KeepAlivePingInterval =
                                               TimeSpan.FromSeconds(10);
                options.Limits.Http2.KeepAlivePingTimeout =
                                               TimeSpan.FromSeconds(1);
            });
            webBuilder.UseStartup<Startup>();
        });
   ```
   <!-- review: KeepAlivePingInterval not found in RC1. Try testing with RC1. See https://github.com/dotnet/aspnetcore/pull/22565/files see C:/Users/riande/source/repos/WebApplication128/WebApplication128 -->

### Containers

> Prior to .NET 5, building and publishing a Dockerfile for an ASP.NET app required pulling the entire .NET Core SDK and the ASP.NET image. With this release, pulling the SDK images bytes is reduced and the bytes pulled for the ASP.NET image is largely eliminated. For more information, See [this GitHub issue comment](https://github.com/dotnet/dotnet-docker/issues/1814#issuecomment-625294750).

Reviewer: ```suggestion
Prior to .NET 5.0, building and publishing a *Dockerfile* for an ASP.NET Core app required pulling the entire .NET Core SDK and the ASP.NET Core image. With this release, pulling the SDK images bytes is reduced and the bytes pulled for the ASP.NET Core image is largely eliminated. For more information, see [this GitHub issue comment](https://github.com/dotnet/dotnet-docker/issues/1814#issuecomment-625294750).
```

### Control Startup class activation

An additional <xref:Microsoft.AspNetCore.Hosting.HostingAbstractionsWebHostBuilderExtensions.UseStartup%2A> overload has been added that lets an app provide a factory method for controlling `Startup` class activation. Controlling `Startup` class activation is useful to pass additional parameters to `Startup` that are initialized along with the host:

```csharp
public class Program
{
    public static async Task Main(string[] args)
    {
        var logger = CreateLogger();
        var host = Host.CreateDefaultBuilder()
            .ConfigureWebHost(builder =>
            {
                builder.UseStartup(context => new Startup(logger));
            })
            .Build();

        await host.RunAsync();
    }
}
```

### Auto refresh with dotnet watch

In .NET 5, running [dotnet watch](xref:tutorials/dotnet-watch) on an ASP.NET Core project both launches the default browser and auto refreshes the browser as changes are made to the code. This means you can:

* Open an ASP.NET Core project in a text editor.

> * Run `dotnet watch`

Reviewer: ```suggestion
* Run `dotnet watch`.
```

* Focus on the code changes while the tooling handles rebuilding, restarting, and reloading the app.

We hope to bring the auto refresh functionality to Visual Studio in the future.

### Console Logger Formatter

Improvements have been made to the console log provider in the `Microsoft.Extensions.Logging` library. Developers can now implement a custom `ConsoleFormatter` to exercise complete control over formatting and colorization of the console output. The formatter APIs allow for rich formatting by implementing a subset of the VT-100 escape sequences. VT-100 is supported by most modern terminals. The console logger can parse out escape sequences on unsupported terminals allowing developers to author a single formatter for all terminals.

### JSON Console Logger

In addition to support for custom formatters, we’ve also added a built-in JSON formatter that emits structured JSON logs to the console. The following code shows how to switch from the default logger to JSON:

[!code-csharp[](~/release-notes/sample/ProgramJsonLog.cs?name=snippet)]

> Log messages emitted to the console are  JSON formatted:

Reviewer: ```suggestion
Log messages emitted to the console are JSON formatted:
```

---
title: What's new in ASP.NET Core 5.0
author: rick-anderson
description: Learn about the new features in ASP.NET Core 5.0.
ms.author: riande
ms.custom: mvc

> ms.date: 12/05/2019

Reviewer: ```suggestion
ms.date: 10/27/2020
```

### Model binding and validation with C# 9 record types

[C# 9 record types](/dotnet/csharp/whats-new/csharp-9#record-types) can be used with model binding in an MVC controller or a Razor Page. Record types are a good way to model data being transmitted over the network.

For example, the following `PersonController` uses the `Person` record type with model binding and form validation:

```csharp
public record Person([Required] string Name, [Range(0, 150)] int Age);

public class PersonController
{
   public IActionResult Index() => View();

   [HttpPost]
   public IActionResult Index(Person person)
   {
          // ...
   }
}
```

The `Person/Index.cshtml` file:

```cshtml
@model Person

Name: <input asp-for="Model.Name" />
<span asp-validation-for="Model.Name" />

Age: <input asp-for="Model.Age" />
<span asp-validation-for="Model.Age" />
```

### Improvements to DynamicRouteValueTransformer

> ASP.NET Core 3.1 introduced <xref:Microsoft.AspNetCore.Mvc.Routing.DynamicRouteValueTransformer> as a way to use use a custom endpoint to dynamically select an MVC controller action or a Razor page. ASP.NET Core 5.0 apps can pass state to a `DynamicRouteValueTransformer` and filter the set of endpoints chosen.

Reviewer 2: ```suggestion
ASP.NET Core 3.1 introduced <xref:Microsoft.AspNetCore.Mvc.Routing.DynamicRouteValueTransformer> as a way to use custom endpoint to dynamically select an MVC controller action or a Razor page. ASP.NET Core 5.0 apps can pass state to a `DynamicRouteValueTransformer` and filter the set of endpoints chosen.
```

### Miscellaneous

> * The <xref:System.ComponentModel.DataAnnotations.CompareAttribute> can be applied to properties on Razor Page model.

Reviewer: ```suggestion
* The [[Compare]](xref:System.ComponentModel.DataAnnotations.CompareAttribute) attribute can be applied to properties on a Razor Page model.
```

* Parameters and properties bound from the body are considered required by default. <!-- Review: How is this different from 3.1
The validation system in .NET Core 3.0 and later treats non-nullable parameters or bound properties as if they had a [Required] attribute.
see https://docs.microsoft.com/aspnet/core/mvc/models/validation?view=aspnetcore-3.1   
-->

## Web API

### OpenAPI Specification on by default

> [OpenAPI Specification](http://spec.openapis.org/oas/v3.0.3) is a industry adopted convention for describing HTTP APIs and integrating them into complex business processes or with 3rd parties. OpenAPI is widely supported by all cloud providers and many API registries. Apps that emit OpenAPI documents from Web APIs have a variety of new opportunities in which those APIs can be used. In partnership with the maintainers of the open-source project [Swashbuckle.AspNetCore](https://www.nuget.org/packages/Swashbuckle.AspNetCore/), the ASP.NET Core API template contains a NuGet dependency on [Swashbuckle](https://github.com/domaindrivendev/Swashbuckle.AspNetCore), a popular open-source NuGet package that emits OpenAPI documents dynamically. Swashbuckle does this by introspecting over the API Controllers and generating the OpenAPI document at run-time, or at build time using the Swashbuckle CLI.

Reviewer: ```suggestion
[OpenAPI Specification](http://spec.openapis.org/oas/v3.0.3) is an industry standard for describing HTTP APIs and integrating them into complex business processes or with third parties. OpenAPI is widely supported by all cloud providers and many API registries. Apps that emit OpenAPI documents from web APIs have a variety of new opportunities in which those APIs can be used. In partnership with the maintainers of the open-source project [Swashbuckle.AspNetCore](https://www.nuget.org/packages/Swashbuckle.AspNetCore/), the ASP.NET Core API template contains a NuGet dependency on [Swashbuckle](https://github.com/domaindrivendev/Swashbuckle.AspNetCore). Swashbuckle is a popular open-source NuGet package that emits OpenAPI documents dynamically. Swashbuckle does this by introspecting over the API controllers and generating the OpenAPI document at run-time, or at build time using the Swashbuckle CLI.
```

In ASP.NET Core 5.0, the web API templates enable the OpenAPI support by default. To disable OpenAPI:

* From the command line:

	```dotnetcli

> 	dotnet new webapi --no-openapi true`

Reviewer: ```suggestion
	dotnet new webapi --no-openapi true
```

	```
* From Visual Studio: Uncheck **Enable OpenAPI support**.

All *.csproj* files created for web API projects contain the [Swashbuckle.AspNetCore](https://www.nuget.org/packages/Swashbuckle.AspNetCore/) NuGet package reference.

```xml
<ItemGroup>
    <PackageReference Include="Swashbuckle.AspNetCore" Version="5.5.1" />
</ItemGroup>
```

The template generated code contains code in `Startup.ConfigureServices` that activates OpenAPI document generation:

[!code-csharp[](~/release-notes/sample/StartupSwagger.cs?name=snippet)]

> The `Configure` method adds the Swashbuckle middleware, which enables the:

Reviewer: ```suggestion
The `Startup.Configure` method adds the Swashbuckle middleware, which enables the:
```

* Document generation process.
* Swagger UI page by default in development mode.

The template generated code won't accidentally expose the API's description when publishing to production.

[!code-csharp[](~/release-notes/sample/StartupSwagger.cs?name=snippet2)]

#### Azure API Management Import

> When ASP.NET Core API projects enable OpenAPI, the Visual Studio 2019 version 16.8 Preview 3.2 and later publishing automatically offer an additional step in the publishing flow. Developers who use [Azure API Management](xref:tutorials/publish-to-azure-api-management-using-vs) have an opportunity to automatically import the APIs into Azure API Management during the publish flow:

Reviewer 2: I don't think we need to refer to VS previews here. VS 168 will have shipped by the time .NET 5 ships.

```suggestion
When ASP.NET Core API projects enable OpenAPI, the Visual Studio 2019 version 16.8 and later publishing automatically offer an additional step in the publishing flow. Developers who use [Azure API Management](xref:tutorials/publish-to-azure-api-management-using-vs) have an opportunity to automatically import the APIs into Azure API Management during the publish flow:
```

![Azure API Management Import VS publishing](~/release-notes/static/publish-to-apim.png)

> ### Better launch Experience for Web API Projects

Reviewer: ```suggestion
### Better launch experience for web API projects
```

> With OpenAPI enabled by default, the app launching experience (F5) for Web API developers significantly improves. With .NET, the Web API template comes pre-configured to load up the Swagger UI page. The Swagger UI page provides both the documentation added for the published API, and enables testing the APIs with a single click.

Reviewer: ```suggestion
With OpenAPI enabled by default, the app launching experience (F5) for web API developers significantly improves. With .NET, the web API template comes pre-configured to load up the Swagger UI page. The Swagger UI page provides both the documentation added for the published API, and enables testing the APIs with a single click.
```

![swagger/index.html view](~/release-notes/static/swagger-ui-page-rc1.png)

## Blazor

### `JSObjectReference` replacement

`JSObjectReference` has been made internal, and `IJSObjectReference` takes its place in the public API. For more information, see the [official announcement (aspnet/Announcements #435)](https://github.com/aspnet/Announcements/issues/435) and <xref:blazor/call-javascript-from-dotnet?view=aspnetcore-5.0#blazor-javascript-isolation-and-object-references>.

### CSS isolation

Blazor now supports defining CSS styles that are scoped to a given component. Component specific CSS styles make it easier to reason about the styles in an app and to avoid unintentional side effects from global styles. <!-- UNCOMMENT CROSS-LINK AFTER https://github.com/dotnet/AspNetCore.Docs/pull/19956 MERGES: For more information, see <xref:blazor/components/css-isolation>.-->

### New `InputFile` component

> The InputFile component allows reading one or more files selected by a user to be upload. For more information, see <xref:blazor/file-uploads>. For more information, see <xref:blazor/file-uploads>.

Reviewer: ```suggestion
The `InputFile` component allows reading one or more files selected by a user to be upload. For more information, see <xref:blazor/file-uploads>.
```

### Custom validation class attributes

Custom validation class names are useful when integrating with CSS frameworks, such as Bootstrap. For more information, see <xref:blazor/forms-validation#custom-validation-class-attributes>.

### IAsyncDisposable support

Blazor components now support the <xref:System.IAsyncDisposable> interface for the asynchronous release of allocated resources.

### JavaScript isolation and object references

Blazor enables JavaScript isolation in standard [JavaScript modules](https://developer.mozilla.org/docs/Web/JavaScript/Guide/Modules). For more information, see <xref:blazor/call-javascript-from-dotnet#blazor-javascript-isolation-and-object-references>.

### Influence the HTML head in Blazor apps

Use the new `Title`, `Link`, and `Meta` components to programmatically set the title of a page and dynamically add links and meta tags to the HTML head in a Blazor app. For more information, see <xref:blazor/fundamentals/additional-scenarios#influence-html-head-tag-elements>.

### Form components support display name

> The following built-in components support display names with the DisplayName parameter:
> 
> * InputDate
> * InputNumber
> * InputSelect

Reviewer: ```suggestion
The following built-in components support display names with the `DisplayName` parameter:

* `InputDate`
* `InputNumber`
* `InputSelect`
```

## gRPC

Many preformance improvements have been made in [gRPC](https://grpc.io/). <!-- For more information, see [this ]
Waiting for https://github.com/aspnet/specs/pull/523
--> 
For more information, see <xref:grpc/index>.

## SignalR

SignalR Hub filters, called Hub pipelines in ASP.NET SignalR, is a feature that allows code to run before and after Hub methods are called. Running code before and after Hub methods are called is similar to how middleware has the ability to run code before and after an HTTP request. Common uses include logging, error handling, and argument validation.

For more information, see [Use hub filters in ASP.NET Core SignalR](xref:signalr/hub-filters).

### SignalR parallel hub invocations

ASP.NET Core SignalR is now capable of handling parallel hub invocations. The default behavior can be changed to allow clients to invoke more than one hub method at a time:

[!code-csharp[](~/release-notes/sample/StartupSignalRhubs.cs?name=snippet)]

### Added Messagepack support in SignalR Java client

> <!-- Review: I can't find NuGet com.microsoft.signalr.messagepack  -->

Reviewer 2: This is for Java, so it's not a NuGet package. https://mvnrepository.com/artifact/com.microsoft.signalr.messagepack

> <!-- Review: I can't find NuGet com.microsoft.signalr.messagepack  -->
> A new package, `com.microsoft.signalr.messagepack`, adds MessagePack support to the SignalR Java client. To use the MessagePack hub protocol, add `.withHubProtocol(new MessagePackHubProtocol())` to the connection builder:

Reviewer: ```suggestion
A new package, [com.microsoft.signalr.messagepack](https://search.maven.org/search?q=g:com.microsoft.signalr.messagepack), adds MessagePack support to the SignalR Java client. To use the MessagePack hub protocol, add `.withHubProtocol(new MessagePackHubProtocol())` to the connection builder:
```

* Support for additional endpoints types in the sockets transport: Adding to the new API introduced in <xref:System.Net.Sockets?displayProperty=nameWithType>, the sockets default transport in [Kestrel](xref:fundamentals/servers/kestrel) allows binding to both existing file handles and Unix domain sockets. Support for binding to existing file handles enables using the existing `Systemd` integration without requiring the `libuv` transport.
* Custom header decoding in Kestrel: Apps can specify which <xref:System.Text.Encoding> to use to interpret incoming headers based on the header name instead of defaulting to `UTF-8`. Set the <xref:System.Net.Http.SocketsHttpHandler.RequestHeaderEncodingSelector> property on <xref:Microsoft.AspNetCore.Server.Kestrel.KestrelServerOptions> to specify which encoding to use:
 
  ```csharp
  public static IHostBuilder CreateHostBuilder(string[] args) =>
    Host.CreateDefaultBuilder(args)
        .ConfigureWebHostDefaults(webBuilder =>
        {
            webBuilder.ConfigureKestrel(options =>
            {
                options.RequestHeaderEncodingSelector = encoding =>
                {

>                     switch (encoding)
>                     {
>                         case "Host":
>                             return System.Text.Encoding.Latin1;
>                         default:
>                             return System.Text.Encoding.UTF8;
>                     }

Reviewer: Let's modernize the code sample to use a C# 8 switch expression.

```suggestion
                          return encoding switch
                          {
                              "Host" => System.Text.Encoding.Latin1,
                              _ => System.Text.Encoding.UTF8,
                          };
```

                };
            });
            webBuilder.UseStartup<Startup>();
        });
  ```

### Kestrel endpoint-specific options via configuration

Support has been added for configuring Kestrel’s endpoint-specific options via [configuration](xref:fundamentals/configuration/index). The endpoint-specific configurations includes the:

> * Http protocols used

Reviewer: ```suggestion
* HTTP protocols used
```

* TLS protocols used
* Certificate selected
* Cient certificate mode

> Configuration allows specifying which certificate is selected based on the specified server name. The server name is part of the Server Name Indication (SNI) extension to the TLS protocol as indicated by the client. Kestrel’s configuration also supports a wildcard prefix in the host name.

Reviewer: ```suggestion
Configuration allows specifying which certificate is selected based on the specified server name. The server name is part of the Server Name Indication (SNI) extension to the TLS protocol as indicated by the client. Kestrel's configuration also supports a wildcard prefix in the host name.
```

The following example shows how to specify endpoint-specific using a configuration file:

```json
{
  "Kestrel": {
    "Endpoints": {
      "EndpointName": {
        "Url": "https://*",
        "Sni": {
          "a.example.org": {
            "Protocols": "Http1AndHttp2",
            "SslProtocols": [ "Tls11", "Tls12"],
            "Certificate": {
              "Path": "testCert.pfx",
              "Password": "testPassword"
            },
            "ClientCertificateMode" : "NoCertificate"
          },
          "*.example.org": {
            "Certificate": {
              "Path": "testCert2.pfx",
              "Password": "testPassword"
            }
          },
          "*": {
            // At least one sub-property needs to exist per
            // SNI section or it cannot be discovered via
            // IConfiguration
            "Protocols": "Http1",
          }
        }
      }
    }
  }
}

> 

Reviewer: ```suggestion
```

For example, the following `PersonController` uses the `Person` record type with model binding and form validation:

```csharp
public record Person([Required] string Name, [Range(0, 150)] int Age);

public class PersonController
{
   public IActionResult Index() => View();

   [HttpPost]
   public IActionResult Index(Person person)
   {
          // ...
   }
}
```

The `Person/Index.cshtml` file:

```cshtml
@model Person

Name: <input asp-for="Model.Name" />
<span asp-validation-for="Model.Name" />

Age: <input asp-for="Model.Age" />
<span asp-validation-for="Model.Age" />
```

### Control Startup class activation

An additional <xref:Microsoft.AspNetCore.Hosting.HostingAbstractionsWebHostBuilderExtensions.UseStartup%2A> overload has been added that lets an app provide a factory method for controlling `Startup` class activation. Controlling `Startup` class activation is useful to pass additional parameters to `Startup` that are initialized along with the host:

```csharp
public class Program
{
    public static async Task Main(string[] args)
    {

>         var logger = CreateLogger();

Author: @Tratcher  who can point me to a test sample that compiles using the new `Startup` overload?

Reviewer 6: https://github.com/dotnet/aspnetcore/commit/113805aba8162ba7bac42943bcb1324801c20f46#diff-23a36744be1739963b7889a34c271f4406b24689afd6cf6329600d7894d01c64R131

Author: @Tratcher  that's great for a test, but now how a customer would call it.

Reviewer 6: What's wrong with the sample you already have?

Author: > What's wrong with the sample you already have?

I'll use that, but it's not complete. I like to have a complete sample with delegates defined.

* Reusable UI components (Razor components)
* Client-side routing
* Component layouts
* Support for dependency injection
* Forms and validation
* Build component libraries with Razor class libraries
* JavaScript interop

For more information, see <xref:blazor/index>.

### Blazor Server

Blazor decouples component rendering logic from how UI updates are applied. Blazor Server provides support for hosting Razor components on the server in an ASP.NET Core app. UI updates are handled over a SignalR connection.

## gRPC

[gRPC](https://grpc.io/):

For more information, see <xref:grpc/index>.

## SignalR

See [Update SignalR code](xref:migration/22-to-30#signalr) for migration instructions. 

## Performance improvements

* HTTP/2:
  * Significant reductions in allocations in the HTTP/2 code path.
  * Support for [HPack dynamic compression](https://tools.ietf.org/html/rfc7541) of HTTP/2 response headers in [Kestrel](xref:fundamentals/servers/kestrel). For more information, see [Header table size](xref:fundamentals/servers/kestrel#header-table-size) and [HPACK: the silent killer (feature) of HTTP/2](https://blog.cloudflare.com/hpack-the-silent-killer-feature-of-http-2/).

> ## Containers

Author: @mthalman please review containers paragraph and suggest better wording/links/etc/

> Prior to .NET 5, building and publishing a Dockerfile ASP.NET app required pulling the .NET Core SDK and the ASP.NET image. With this release, the SDK images size is reduced and the ASP.NET image is eliminated, only the small manifest needs to be pulled. For more information, see [this GitHub issue comment](https://github.com/dotnet/dotnet-docker/issues/1814#issuecomment-625294750).

Reviewer 7: This could be misinterpreted.  The ASP.NET image hasn't been eliminated.  And today you still end up pulling both the SDK and ASP.NET image.  It's just that they layers contained in the ASP.NET image are the same layers that exist in the SDK image.  So by pulling the SDK image, you've already downloaded all the layers you need for the ASP.NET image.  Which means that pulling the ASP.NET image after having already pulled the SDK image, it ends up not downloading any additional layers.  I wouldn't mention anything about the manifest; that's too much in the weeds.  Just say that no additional layers would end up being downloaded.  Before this change, there was no layer sharing so pulling the ASP.NET image after pulling the SDK image _did_ result in additional layers being downloaded.  This should be characterized as an optimization that was made in 5.0.
